[Merged by Bors] - feat(data/equiv): Add lemmas to reduce `@finset.univ (perm (fin n.succ)) _`

eric-wieser opened the pull request:
The culmination of these lemmas is that `matrix.det` can now be reduced by a minimally steered simp:
```lean
import data.matrix.notation
import group_theory.perm.fin
import linear_algebra.determinant
open finset

example {α : Type} [comm_ring α] {a b c d : α} :
  matrix.det ![![a, b], ![c, d]] = a * d - b * c :=
begin
  simp [matrix.det, univ_perm_fin_succ, ←univ_product_univ, sum_product, fin.sum_univ_succ, fin.prod_univ_succ],
  ring
end
```
---
<!--
put comments you want to keep out of the PR commit here.
If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->
- [x] depends on: #5614 
- [x] depends on: #5622 
- [x] depends on: #5735 
- [x] depends on: #5735 
- [x] depends on: #5736 
- [x] depends on: #5737 
- [x] depends on: #5738  
- [x] depends on: #5739  
- [x] depends on: #5740  
- [x] depends on: #5741

eric-wieser committed:
chore(data/equiv/fin): Add missing simp lemmas

eric-wieser committed:
feat(data/equiv/option): Add equiv.remove_none

eric-wieser committed:
feat(data/equiv/option): Fix redundant argument

eric-wieser committed:
feat(group_theory/perm/option): Show `perm (fin n.succ) ≃ fin n.succ × perm (fin n)`

pechersky committed:
Fix fin size error in proposition

pechersky committed:
generalize equiv_functor.map_equiv.injective

pechersky committed:
Fix docstring to mention fin instead of option

pechersky committed:
add subsingleton instances for equivs

pechersky committed:
refl lemmas for `equiv_congr`

pechersky committed:
mark `swap_self` as simp

pechersky committed:
remove unneeded arg for fin_succ_equiv_symm_none

pechersky committed:
simplify subsingleton perms to refl

pechersky committed:
perm_congr apply lemma of symm

pechersky committed:
remove extra line

pechersky committed:
add copywrite header

pechersky committed:
simplifying lemmas of decompose_fin when on refls

pechersky committed:
fin_succ_equiv apply lemma which simps nicer than succ

pechersky committed:
satisfy linter by remove subsingleton arg

pechersky committed:
satisfy linter by removing fintype constraint

pechersky committed:
add docstrings to equiv.perm.decompose_*

pechersky committed:
remove fintype constraint

pechersky committed:
simp subsingleton perms when they're coed

what would be the most correct way of getting the simplifier to simplify
subsingletons? the linter complained about the simple equality
so instead we simplify when they're coerced

pechersky committed:
simps on decompose_option

eric-wieser commented on src/data/equiv/fin.lean:
```diff
@@ -44,6 +44,25 @@ def fin_succ_equiv (n : ℕ) : fin n.succ ≃ option (fin n) :=
+begin
+  set m' : fin n := m.pred h,
+  rw [show m = (m.pred h).succ, by rw [fin.succ_pred], fin_succ_equiv_succ]
+end
```
This one seems like an odd simp lemma. Did we actually need it to simplify the determinant expression?

pechersky replied:
Without it, the simp expression in the proof of `equiv.perm.decompose_fin_symm_of_refl` won't fire. If you prefer, this can be moved into the body of that proof.

eric-wieser replied:
This can now be removed I think, although I don't want to touch this file for fear of recompilation until we're done trying to reduce 3x3 matrices

eric-wieser commented on src/group_theory/perm/option.lean:
```diff
@@ -0,0 +1,84 @@
+begin
+  ext x,
+  by_cases hp : p = 0;
+  by_cases hx : x = 0;
```
Does the proof become easier if you use `cases hp : p` instead of `by_cases hp : p = 0` here ad similarly for `x`? I think that will remove the need for the lemma above

pechersky replied:
In this case, it's a bit more complicated because `x` and `p` are both `fin (n + 1)` so a simple `cases` just deconstructs the subtype into the value and property witness. Using `refine fin.cases _ _ x` also is not great because the `hx' : x = p` or the inequality is "forgotten" -- the second goal does not relate the now universally quantified `i : fin n` back to `x`.

pechersky replied:
But I think there are some ways to shorten it -- working on it

pechersky replied:
```
@[simp] lemma equiv.perm.decompose_fin_symm_of_refl {n : ℕ} (p : fin (n + 1)) :
  equiv.perm.decompose_fin.symm (p, equiv.refl _) = swap 0 p :=
begin
  ext x,
  refine fin.cases _ _ p;
  refine fin.cases _ _ x;
  simp [equiv.perm.decompose_fin],
  intros x₁ p₁,
  by_cases h : x₁ = p₁;
  simp [h, equiv.perm.decompose_fin, swap_apply_of_ne_of_ne, fin.succ_ne_zero]
end
```
Is that preferred?

eric-wieser replied:
That looks reasonable for me - I'm waiting for CI to generate a cache for me so that I can step through it.

eric-wieser replied:
Having played around with this - ideally we can get rid of the `ext`, and have simp handle it all. This lemma helps:
```lean
@[simp] lemma map_equiv_refl {α : Type*} : equiv_functor.map_equiv option (1 : perm α) = 1 := begin
  ext, simp [equiv_functor.map_equiv, equiv_functor.map],
end
```
as does normalizing with `simp [←equiv.perm.one_def, ←equiv.perm.inv_def]`

eric-wieser replied:
I've gone ahead and pushed a commit with these new lemmas, which reduces this proof to `simp [equiv.perm.decompose_fin]`.

eric-wieser committed:
chore(group_theory/perm/basic): Add some missing simp lemmas

`simp` can't find the appropriate `equiv` lemmas as they are about `refl` not `1`, even though those are defeq.

eric-wieser committed:
Update src/group_theory/perm/basic.lean

eric-wieser committed:
chore(*): Add the determinant test

eric-wieser committed:
chore(*): Split out a new `group_theory.perm.fin` file

pechersky committed:
satisfy simp linter by keeping only perm_congr_apply as simp

pechersky committed:
shorten prefix of term

pechersky committed:
simp lemma to simplify heads of vectors

pechersky committed:
refl, symm, and trans of equiv_functor.map_equiv

pechersky committed:
final parallel defn lemmas for sign

pechersky committed:
simp tag sign_symm_trans_trans, symm version of it

pechersky committed:
equiv_functor option lemmas

pechersky committed:
decompose_option_symm none apply and sign

pechersky committed:
fix imports

pechersky committed:
decompose_fin_symm apply and sign lemmas

pechersky committed:
det of 3x3 matrix example

needs swap simp lemmas written

eric-wieser committed:
Merge branch 'master' into eric-wieser/fin-determinant-simp

pechersky committed:
move the perm_congr lemmas into data/equiv/basic

pechersky committed:
Merge branch 'eric-wieser/fin-determinant-simp' of https://github.com/leanprover-community/mathlib into eric-wieser/fin-determinant-simp

pechersky committed:
remove simp from sign_perm_congr since perm_congr is simped now

pechersky committed:
remove tautological fin_succ_equiv_ne_zero lemma

eric-wieser commented on src/control/equiv_functor.lean:
```diff
@@ -54,6 +54,24 @@ def map_equiv :
+
+lemma map_equiv_symm :
+  (map_equiv f e).symm = map_equiv f e.symm :=
+equiv.ext $ map_equiv_symm_apply f e
```
I think `simp` should be swapped here:
```suggestion
lemma map_equiv_symm_apply (y : f β) :
  (map_equiv f e).symm y = equiv_functor.map e.symm y := rfl

@[simp] lemma map_equiv_refl (α) :
  map_equiv f (equiv.refl α) = equiv.refl (f α) :=
by simpa [equiv_functor.map_equiv]

@[simp] lemma map_equiv_symm :
  (map_equiv f e).symm = map_equiv f e.symm :=
equiv.ext $ map_equiv_symm_apply f e
```

eric-wieser commented on src/data/equiv/basic.lean:
```diff
@@ -256,16 +269,31 @@ def equiv_congr {δ} (ab : α ≃ β) (cd : γ ≃ δ) : (α ≃ γ) ≃ (β ≃
 
-@[simp] lemma perm_congr_apply {α β : Type*} (e : α ≃ β) (p : equiv.perm α) (x) :
+@[simp] lemma perm_congr_def {α β : Type*} (e : α ≃ β) (p : equiv.perm α) :
+  e.perm_congr p = (e.symm.trans p).trans e := rfl
```
I'm not certain that this is a good simp lemma, although it's certainly a useful lemma.

If we're going to add this one, we should also add a similar one for `equiv_congr_def`, which would make `equiv_congr_refl_left` and `equiv_congr_refl_right` unecessary.

pechersky replied:
I'm not sure how to express `equiv_congr_def` in a clean way.

pechersky committed:
simp on the unapplied map_equiv_symm instead

pechersky committed:
chore(group_theory/perm/sign): remove classical for sign congr simp lemmas

pechersky committed:
Merge branch 'master' into eric-wieser/fin-determinant-simp

pechersky committed:
Merge branch 'pechersky/perm-sign-congr-simp' into eric-wieser/fin-determinant-simp

pechersky committed:
prepare to shuffle simp tags in basic

pechersky committed:
shorten determinant proof relying on sign simps

pechersky committed:
Merge branch 'pechersky/perm-sign-congr-simp' into eric-wieser/fin-determinant-simp

pechersky committed:
Merge remote-tracking branch 'origin/eric-wieser/one-vs-refl' into eric-wieser/fin-determinant-simp

pechersky committed:
remove gh-5614 duplicates

pechersky committed:
shuffle simp tags of perm_congr lemmas

pechersky commented on src/data/equiv/basic.lean:
```diff
@@ -256,16 +269,31 @@ def equiv_congr {δ} (ab : α ≃ β) (cd : γ ≃ δ) : (α ≃ γ) ≃ (β ≃
 @[simp] lemma perm_congr_apply {α β : Type*} (e : α ≃ β) (p : equiv.perm α) (x) :
 e.perm_congr p x = e (p (e.symm x)) := rfl
 
+@[simp] lemma perm_congr_symm_apply {α β : Type*} (e : α ≃ β) (p : equiv.perm β) (x) :
```
```suggestion
lemma perm_congr_symm_apply {α β : Type*} (e : α ≃ β) (p : equiv.perm β) (x) :
```

pechersky committed:
satisfy linter

eric-wieser committed:
Merge branch 'master' into eric-wieser/fin-determinant-simp

eric-wieser commented on src/data/equiv/basic.lean:
```diff
@@ -301,6 +301,9 @@ lemma perm_congr_def {α β : Type*} (e : α ≃ β) (p : equiv.perm α) :
 
+lemma perm_congr_symm_apply {α β : Type*} (e : α ≃ β) (p : equiv.perm β) (x) :
+e.perm_congr.symm p x = e.symm (p (e x)) := rfl
+
```
```suggestion
```

eric-wieser committed:
Update src/data/equiv/basic.lean

kbuzzard commented on test/matrix.lean:
```diff
@@ -50,4 +52,25 @@ example {a b c d e f g h : α} : ![a, b, c, d, e, f, g, h] 7 = h := by simp
+  simp [h0, h1, hA, matrix.det, finset.univ_perm_fin_succ, ←finset.univ_product_univ,
+        finset.sum_product, fin.sum_univ_succ, fin.prod_univ_succ, fin.succ_ne_zero],
+  ring
+end
```
These proofs are not things that someone like me is going to be able to find themselves, and nobody will think to look in `test`. Why not just put them in mathlib?

eric-wieser replied:
The idea behind these proofs was:
* To remind ourselves that there is more work to be done here - ideally this would be closed with `simp` or perhaps `simp with <some simp set>`
* To ensure that other PRs don't undo the progress made here

If the `matrix.det ![![a, b], ![c, d]] = a * d - b * c` lemma was directly in mathlib, it would probably need to have the form
`matrix.det A = A 0 0 * A 1 1 - A 0 1 * A 1 0`.

But I think the idea was to leave this type of thing to a follow-up PR.

pechersky replied:
A simp set can be constructed that is meant to unwrap sums as deep as possible. A lot of lemma would need to be tagged though for that simp set.

pechersky committed:
Merge branch 'master' into eric-wieser/fin-determinant-simp

pechersky committed:
utilize norm_num with norm_swap plugin

Vierkantor reviewed (commented):
The code itself looks good, though I feel the documentation is a bit lacking. I understand why we would want the results not to be marked `simp`, but that makes them hard to find as Kevin pointed out. Perhaps you can point out the examples `test/matrix.lean` in either `matrix/notation.lean` or `determinant.lean`?

pechersky committed:
Standardize typing in det examples

pechersky committed:
Refer to examples in data/matrix/notation

Vierkantor committed:
Module doc for `matrix/determinant.lean`

Vierkantor commented on src/data/matrix/notation.lean:
```diff
@@ -34,6 +34,10 @@ already appears in the input.
+
+## Examples
+
+Examples of usage can be found in the `test/matrix` file.
```
```suggestion
Examples of usage can be found in the `test/matrix.lean` file.
```

Vierkantor committed:
`test/matrix` -> `test/matrix.lean`

Vierkantor reviewed (approved):
The changes look good to me, but I wrote a little bit myself. @jcommelin, could you do the final(?) review?

jcommelin commented on src/data/equiv/option.lean:
```diff
@@ -0,0 +1,72 @@
+import control.equiv_functor
+
+/-!
+# Equivalences for `option α`
```
Note that naively I would expect to find in this file a definition that turns `equiv X Y` into `equiv (option X) (option Y)`.
Surely this somewhere else in mathlib. Could you please add a pointer to it in this module docstring?

pechersky replied:
Added

jcommelin commented on src/data/equiv/option.lean:
```diff
@@ -0,0 +1,72 @@
+end
+
+/-- Given an equivalence between two `option` types, eliminate `none` from that equivalence by
+mapping `e.symm none` to `e none`. -/
```
Should there be a simp-lemma that witnesses this fact?

pechersky replied:
One related simp lemma (`option_symm_apply_none_iff`) was added. The lemma that witnesses the docstring fact `some_remove_none_iff` is an iff that does not seem helpful to tag as simp. If that's not the case, let me know.

jcommelin reviewed (approved):
Thanks! LGTM

two minor comments, otherwise let's merge this

pechersky committed:
add docstring

pechersky committed:
add lemmas that match remove_none docstring
